File: shared/sensors.py
def parseSensors(data: list, sysInfo: dict):
    return {
        'batt_charge': data.pop(0),
        'batt_level': data.pop(0),
        'batt_time': data.pop(0),
        'bclk': data.pop(0),
        'cpu_fan': data.pop(0),
        'cpu_power': data.pop(0),
        'cpu_temp': data.pop(0),
        'cpu_usage': data.pop(0),
        'ram_freq': data.pop(0),
        'ram_used': data.pop(0),
        'cpus': [data.pop(0) for _ in range(sysInfo['cpuThreads'])],
        'drives': [{
            'read': data.pop(0),
            'write': data.pop(0)
        } for _ in range(len(sysInfo['drives']))],
        'smart': [{
            'temp': data.pop(0),
            'life': data.pop(0),
            'warning': data.pop(0),
            'failure': data.pop(0),
            'reads': data.pop(0),
            'writes': data.pop(0)
        } for _ in range(len(sysInfo['drives']))],
        'networkInterfaces': [{
            'up': data.pop(0),
            'dl': data.pop(0)
        } for _ in range(len(sysInfo['networkInterfaces']))],
        'gpus': [{
            'usage': data.pop(0),
            'mem_used': data.pop(0),
            'temp': data.pop(0)
        } for _ in range(len(sysInfo['gpus']))]
    }

File: shared-ts/src/sensors.ts
import { Sensors, SysInfo } from './interfaces';
import { range } from './utils';


export const parseSensors = (sysInfo: SysInfo, data: number[]): Sensors => {
  return {
    batt_charge: data.shift() ?? -1,
    batt_level: data.shift() ?? -1,
    batt_time: data.shift() ?? -1,
    bclk: data.shift() ?? -1,
    cpu_fan: data.shift() ?? -1,
    cpu_power: data.shift() ?? -1,
    cpu_temp: data.shift() ?? -1,
    cpu_usage: data.shift() ?? -1,
    ram_freq: data.shift() ?? -1,
    ram_used: data.shift() ?? -1,
    cpus: range(sysInfo.cpuThreads).map(() => data.shift() ?? -1),
    drives: sysInfo.drives.map(() => ({
      read: data.shift() ?? -1,
      write: data.shift() ?? -1,
    })),
    smart: sysInfo.drives.map(() => ({
      temp: data.shift() ?? -1,
      life: data.shift() ?? -1,
      warning: data.shift() ?? -1,
      failure: data.shift() ?? -1,
      reads: data.shift() ?? -1,
      writes: data.shift() ?? -1,
    })),
    networkInterfaces: sysInfo.networkInterfaces.map(() => ({
      up: data.shift() ?? -1,
      dl: data.shift() ?? -1,
    })),
    gpus: sysInfo.gpus.map(() => ({
      usage: data.shift() ?? -1,
      mem_used: data.shift() ?? -1,
      temp: data.shift() ?? -1,
    })),
  };
};
